### PR TITLE
SONARKT-89 S6307: Suspending functions should be main-safe

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**

### CODE REVIEW

File: kotlin-checks-test-sources/src/main/kotlin/checks/MainSafeCoroutinesCheckSample.kt [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        withContext(Dispatchers.IO) {
+            f()
+        }
+        delay(500L)
+    }
+}
+
+@Throws(IOException::class)
+fun f() {
+}
```
Comment (johann-beleites-sonarsource) [inline kotlin-checks-test-sources/src/main/kotlin/checks/MainSafeCoroutinesCheckSample.kt:116]: How about a compliant `Thread.sleep` in a non-suspending fun, to test we don't have FPs there?
Same for a call to `f()`?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        }
+    }
+
+    private fun KtElement.reportBlockingFunctionCalls(context: KotlinFileContext) {
+        val bindingContext = context.bindingContext
+        forEachDescendantOfType<KtCallExpression> { call ->
+            val resolvedCall = call.getResolvedCallNative(bindingContext)
+            if (resolvedCall matches THREAD_SLEEP_MATCHER) {
+                context.reportIssue(call.calleeExpression!!,
+                    """Replace this "Thread.sleep()" call with "delay()"""")
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt:86]: `\n` doesn't seem necessary here

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            } else {
+                resolvedCall?.resultingDescriptor?.let { descriptor ->
+                    if (isInsideNonSafeDispatcher(call,
+                            bindingContext) && descriptor.annotations.any { throwsBlockingExceptions(it) }
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt:90]: Odd `\n` here

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+private fun throwsBlockingExceptions(annotation: AnnotationDescriptor) =
+    (annotation.fqName?.asString() == THROWS_FQN
+        && (annotation.allValueArguments.asSequence()
+        .find { (k, _) -> k.asString() == "exceptionClasses" }
+        ?.value as? ArrayValue)
+        ?.value
+        ?.mapNotNull { it.value as? KClassValue.Value.NormalClass }
+        ?.map { it.value.toString().replace("/", ".") }
+        ?.any(BLOCKING_ANNOTATIONS::contains)
+        ?: false)
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt:144]: Looks like you can get rid of the first and final `(` and `)`. Also a suggestion for formatting.
```suggestion
    annotation.fqName?.asString() == THROWS_FQN &&
        (annotation.allValueArguments.asSequence()
            .find { (k, _) -> k.asString() == "exceptionClasses" }
            ?.value as? ArrayValue)
            ?.value
            ?.mapNotNull { it.value as? KClassValue.Value.NormalClass }
            ?.map { it.value.toString().replace("/", ".") }
            ?.any(BLOCKING_ANNOTATIONS::contains)
        ?: false
```

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+private fun KtLambdaArgument.isSuspending(
+    bindingContext: BindingContext,
+) = (parent as? KtCallExpression)
+    ?.getResolvedCallNative(bindingContext)
+    ?.resultingDescriptor
+    ?.valueParameters
+    ?.last()
+    ?.hasSuspendFunctionType
+    ?: false
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt:154]: Perhaps a method for `ApiExtensions`?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        ?.fqNameOrNull()
+        ?.asString()
+
+    return arg == null
+        || arg is DefaultValueArgument
+        || argValue == "$KOTLINX_COROUTINES_PACKAGE.Dispatchers.Main"
+        || argValue == "$KOTLINX_COROUTINES_PACKAGE.Dispatchers.Default"
+}
+
+private fun throwsBlockingExceptions(annotation: AnnotationDescriptor) =
```
Comment (johann-beleites-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt:135]: Could this be valuable as function in `ApiExtensions`? Which takes e.g. a collection of exceptions that it checks for? That way we can centrally update it as well, once we find / the compiler provides us with a better way to detect which exceptions may be thrown.

**Review by johann-beleites-sonarsource — COMMENTED** [review]
A couple of comments, the majority minor

File: kotlin-checks-test-sources/src/main/kotlin/checks/MainSafeCoroutinesCheckSample.kt (hunk 1 above)
Comment (margarita-nedzelska-sonarsource) [reply kotlin-checks-test-sources/src/main/kotlin/checks/MainSafeCoroutinesCheckSample.kt:116]: Is it what you're talking about? [L79-L81](https://github.com/SonarSource/sonar-kotlin/pull/28/files/664bb3413d9500ff73c4f7b37f274870518f541b#diff-76cc128772b9466def676256139e9ebd35c885f751dc6230da1ac46003ad6a43R79-R81)

Comment (johann-beleites-sonarsource) [reply kotlin-checks-test-sources/src/main/kotlin/checks/MainSafeCoroutinesCheckSample.kt:116]: My bad, yes!

**Review by johann-beleites-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**